### PR TITLE
Updates regarding SMB shares

### PR DESCRIPTION
**Pull request by anodos325**
In addition to these, enabling NTLMv1 auth is now handled by an checkbox in the UI under Services->SMB. No need to add an aux parameter.

### CODE REVIEW

**Comment by anodos325** [comment]
Additional information about commit ab7d809 (Clarifications and corrections regarding SMB):

1.  When SMB config is changed through Services->SMB, the samba server is restarted.
2.  When share config is changed through Sharing->Windows (SMB) Shares-><share name>, the smb4.conf file is reloaded. 

(2) does not affect existing sessions while (1) does affect them. It's important to know what to expect on the back-end when making UI changes. ;-)

File: userguide/services.rst [hunk 1]
```diff
@@ -1635,7 +1627,9 @@ unless there is a specific need.**
   back up the data, create a new case-insensitive dataset, create an
   SMB share on it, then copy the data from the old one onto it. After
   the data has been checked and verified on the new share, the old one
-  can be deleted.
+  can be deleted. Once the case-insensitive dataset is created, the
```
Comment (wblock) [inline userguide/services.rst]: This seems to be in the wrong place.  Setting the parameter should be done right after creating the new share, before copying data.

Also, the "Once" construct can be confusing to non-native English speakers.  "When" or "After" is somewhat clearer.

File: userguide/sharing.rst [hunk 2]
```diff
@@ -1057,9 +1057,9 @@ provides more details for each configurable option.
...
    |                              |               |          |                                                                                                             |
    +------------------------------+---------------+----------+-------------------------------------------------------------------------------------------------------------+
-   | Allow Guest Access           | checkbox      |          | if checked, no password is required to connect to the share and all users share the permissions of the      |
-   |                              |               |          | guest user defined in the :ref:`SMB` service                                                                |
-   |                              |               |          |                                                                                                             |
+   | Allow Guest Access           | checkbox      |          | if checked, no password is required to connect to the share. User login attempts with a bad password are    |
```
Comment (wblock) [inline userguide/sharing.rst]: See note above about sentences in tables.  "No password is required" is less clear than "a password is not required".  "User login attempts" is not clear. A user does not log in to a share, they connect to it.

File: userguide/sharing.rst [hunk 3]
```diff
@@ -1092,16 +1092,17 @@ settings:
...
-  checked (the default), other users will see the names of every share
-  that exists using Windows Explorer, but they will receive a
-  permissions denied error message if they try to access someone
-  else's share. If this box is unchecked, even the owner of the share
-  will not see it or be able to create a drive mapping for the share
-  in Windows Explorer. However, they can still access the share from
-  the command line. Unchecking this option provides limited security
-  and is not a substitute for proper permissions and password control.
+* When the :guilabel:`Browsable to Network Clients` box is checked (the 
+  default), the share will be visible in the server's browse list. If the share
```
Comment (wblock) [inline userguide/sharing.rst]: "visible in the server's browse list" is not clear.  Also, use "is" instead of "will be".

File: userguide/sharing.rst [hunk 4]
```diff
@@ -1092,16 +1092,17 @@ settings:
...
-  that exists using Windows Explorer, but they will receive a
-  permissions denied error message if they try to access someone
-  else's share. If this box is unchecked, even the owner of the share
-  will not see it or be able to create a drive mapping for the share
-  in Windows Explorer. However, they can still access the share from
-  the command line. Unchecking this option provides limited security
-  and is not a substitute for proper permissions and password control.
+* When the :guilabel:`Browsable to Network Clients` box is checked (the 
+  default), the share will be visible in the server's browse list. If the share 
+  is is flagged as a :guilabel:`home share`, unchecking this box will hide the
```
Comment (wblock) [inline userguide/sharing.rst]: Duplicate "is" at start of line.
Also, s/will hide/hides/

File: userguide/sharing.rst [hunk 5]
```diff
@@ -1092,16 +1092,17 @@ settings:
...
-  permissions denied error message if they try to access someone
-  else's share. If this box is unchecked, even the owner of the share
-  will not see it or be able to create a drive mapping for the share
-  in Windows Explorer. However, they can still access the share from
-  the command line. Unchecking this option provides limited security
-  and is not a substitute for proper permissions and password control.
+* When the :guilabel:`Browsable to Network Clients` box is checked (the 
+  default), the share will be visible in the server's browse list. If the share 
+  is is flagged as a :guilabel:`home share`, unchecking this box will hide the 
+  [homes] share so that only the dynamically generated home share with the
```
Comment (wblock) [inline userguide/sharing.rst]: Markup seems needed for [homes], but it's not clear to me what is being denoted there.  Literal value, variable value, setting in configuration file...

File: userguide/sharing.rst [hunk 6]
```diff
@@ -1092,16 +1092,17 @@ settings:
...
+  authenticated user's name will be visible (in contrast with [homes]
```
Comment (wblock) [inline userguide/sharing.rst]: In general, avoid possessives.  It is usually not needed: "username" rather than "user's name", for example.

File: userguide/sharing.rst [hunk 7]
```diff
@@ -1092,16 +1092,17 @@ settings:
...
+  and [<username>] both being visible).
```
Comment (wblock) [inline userguide/sharing.rst]: Again, markup is needed.

File: userguide/sharing.rst [hunk 8]
```diff
@@ -1092,16 +1092,17 @@ settings:
...
-  the command line. Unchecking this option provides limited security
-  and is not a substitute for proper permissions and password control.
+* When the :guilabel:`Browsable to Network Clients` box is checked (the 
+  default), the share will be visible in the server's browse list. If the share 
+  is is flagged as a :guilabel:`home share`, unchecking this box will hide the 
+  [homes] share so that only the dynamically generated home share with the
+  authenticated user's name will be visible (in contrast with [homes]
+  and [<username>] both being visible). 
+
+  The presence of a share in a user's browse list does not grant the user
```
Comment (wblock) [inline userguide/sharing.rst]: Hard to tell what you mean by "a user's browse list".  Does this mean "Shares that appear in the File Explorer" (or whatever Windows calls it today) "do not necessarily mean the user has read or write permissions to them."?

File: userguide/sharing.rst [hunk 9]
```diff
@@ -1092,16 +1092,17 @@ settings:
...
+  default), the share will be visible in the server's browse list. If the share 
+  is is flagged as a :guilabel:`home share`, unchecking this box will hide the 
+  [homes] share so that only the dynamically generated home share with the
+  authenticated user's name will be visible (in contrast with [homes]
+  and [<username>] both being visible). 
+
+  The presence of a share in a user's browse list does not grant the user  
+  read or write permissions on the share. Users are still able to directly 
+  access and map shares that are not visible in the browse list by using 
+  the share's UNC path. Hence, this option is of limited security utility.
```
Comment (wblock) [inline userguide/sharing.rst]: "This option provides no real security because shares that are not visible in the File Explorer (or whatever) can still be accessed with a *UNC* path."

Not sure about markup, but I think that's what this is trying to say.

File: userguide/sharing.rst [hunk 10]
```diff
@@ -1133,13 +1134,7 @@ like Windows 7 will not be able to connect with NTLMv1 disabled.
...
 <https://support.microsoft.com/en-us/help/2793313/security-guidance-for-ntlmv1-and-lm-network-authentication>`_
 has information about the security implications and ways to enable
-NTLMv2. If changing the client configuration is not possible, NTLMv1
-authentication can be enabled by adding this entry to
-:guilabel:`Auxiliary Parameters`:
-
-.. code-block:: none
-
-   ntlm auth = yes
+NTLMv2.
```
Comment (wblock) [inline userguide/sharing.rst]: This removes the mention of NTLMv1 entirely.  Shouldn't it point to the checkmark option?

File: userguide/sharing.rst [hunk 11]
```diff
@@ -1292,27 +1287,27 @@ for more details.
    | xattr_tdb           | stores Extended Attributes (EAs) in a tdb file so they can be used on filesystems which do not provide support for EAs                     |
    |                     |                                                                                                                                            |
    +---------------------+--------------------------------------------------------------------------------------------------------------------------------------------+
-
+   | zfs_space           | correctly calculates ZFS space used by share, including space used by ZFS snapshots, quotas, and resevations. Enabled by default.          |
```
Comment (wblock) [inline userguide/sharing.rst]: No full sentences in table descriptions.

File: userguide/sharing.rst [hunk 12]
```diff
@@ -1292,27 +1287,27 @@ for more details.
...
+   |                     |                                                                                                                                            |
+   +---------------------+--------------------------------------------------------------------------------------------------------------------------------------------+
+   | zfsacl              | provide ACL extensions for proper integration with ZFS. Enabled by default.                                                                |
```
Comment (wblock) [inline userguide/sharing.rst]: As above, no full sentences.  (I don't know why, it was that way when I got here.)

File: userguide/sharing.rst [hunk 13]
```diff
@@ -1292,27 +1287,27 @@ for more details.
...
 
 .. note:: When using :guilabel:`fruit`, also add the
    :guilabel:`streams_xattr` and :guilabel:`catia` VFS objects and be
    sure to configure **all** SMB shares this way. Reboot the Mac
    client after making this change.
 
-These VFS objects do not appear in the drop-down menu as they are
-always enabled:
+These VFS objects do not appear in the drop-down menu and are dynamically
+configured:
```
Comment (wblock) [inline userguide/sharing.rst]: What does "dynamically configured" mean?  Set by the system, set by the user at runtime?

File: userguide/sharing.rst [hunk 14]
```diff
@@ -1292,27 +1287,27 @@ for more details.
...
 
 * **recycle:** moves deleted files to the recycle directory instead of
-  deleting them
+  deleting them. Controlled by a checkbox in the share configuration.
```
Comment (wblock) [inline userguide/sharing.rst]: Second sentence needs to use :guibutton: or a :ref: to the appropriate section.

File: userguide/services.rst [hunk 15]
```diff
@@ -1484,7 +1484,7 @@ This configuration screen is really a front-end to
    | SMB2_10        | used by Windows 7                                          |
    |                |                                                            |
    +----------------+------------------------------------------------------------+
-   | SMB3           | used by Windows 8                                          |
+   | SMB3           | Same as SMB3_11                                            |
```
Comment (anodos325) [inline userguide/services.rst]: I see your point. Since SMB3 = SMB3_11, "used by Windows 8" is technically incorrect. This is relevant when setting server minimum protocol = SMB3.  
Does the following look better?
**used by Windows 10; same as SMB3_11**

File: userguide/services.rst [hunk 16]
```diff
@@ -1537,15 +1538,6 @@ a multiple of _SC_PAGESIZE (typically *4096*) to avoid memory
...
 
-To increase network performance, read the Samba section on socket
-options in the
-`smb.conf manual page
-<https://www.freebsd.org/cgi/man.cgi?query=smb.conf&manpath=FreeBSD+11.0-RELEASE+and+Ports>`_.
-It indicates which options are available and recommends that you
-experiment to see which are supported by your clients and improve your
-network's performance.
-#endif freenas
-
```
Comment (anodos325) [inline userguide/services.rst]: Heh. I'm kind of new to this git/github thing and didn't realize any additional edits I make get appended to existing pull requests. :-)

Adjusting socket options is somewhat frowned upon in the samba mailing lists (JRA has even mentioned ripping the parameter out of samba entirely). Most edits people do are voodoo, and some of them actually hurt performance. It's probably better to adjust send and receive buffers through sysctls (which people with Intel 10GB NICs tend to do). This is because it will still allow the kernel to automatically tune send and receive buffers. 

Manipulating IPTOS headers may help some people who have samba passing through routers (and have configured QOS in routers accordingly), but I believe this probably is a more advanced topic than most people looking at the FreeNAS handbook are trying to address.

TL;DR, socket options aren't exactly where I'd point people to if they were trying to improve samba performance.

I'll fix the #endif freenas. I'm still learning the ropes of the markup language. I'll also leave the blurb about socket options if you want ;-)

File: userguide/services.rst (hunk 1 above)
Comment (anodos325) [inline userguide/services.rst]: My writing can also confuse native English speakers. :) The problem with the way that it is written originally is that no performance improvement will occur until the admin disables the case sensitivity checks. 

Another way of looking at it is that case sensitivity checks cause o(n^2) delay. Disabling them speeds things up. Making the dataset case insensitive allows you to do this safely. 

I'm open to better ways to say this.

File: userguide/sharing.rst (hunk 5 above)
Comment (anodos325) [inline userguide/sharing.rst]: Sorry. This was half-baked. [homes] is a special section of the smb.conf file that is generated when when the user checks the box "Use as home share". Enabling home shares in samba in FreeNAS will basically generate two shares:
\\freenas\homes
\\freenas\%U
Unchecking "browsable" for a home share makes it so that \\freenas\homes isn't visible, but \\freenas\%U will be visible.

File: userguide/sharing.rst (hunk 13 above)
Comment (anodos325) [inline userguide/sharing.rst]: In my defense, it made sense when I was writing it at 4:30 AM before drinking coffee.

File: userguide/sharing.rst (hunk 14 above)
Comment (b-a-t) [inline userguide/sharing.rst]: If you dive into the protocol version differences, it's worth mentioning, that most of the time when you talk about SMB2, you should mean SMB2_02. See [#24334](https://bugs.freenas.org/issues/24334).

>     SMB1 -> NT1
>     SMB2 -> SMB2_02
>     SMB2.1 -> SMB2_10
>     SMB3 -> SMB3_00

I hope to fix that in a GUI one day, around 11.1, I hope, but meanwhile clarifying it in the docs could be an option.

Comment (anodos325) [inline userguide/sharing.rst]: I believe that **server min protocol = SMB2** is identical to **server min protocol  = SMB2_10**
This means that Vista clients will fail to negotiate a supported protocol. Yay for inconsistency! 

Perhaps we need a errata section (or separate book) for weird SMB client behavior.

File: userguide/sharing.rst [hunk 17]
```diff
@@ -1093,16 +1093,15 @@ settings:
...
-  [homes] share so that only the dynamically generated home share with the
-  authenticated user's name will be visible (in contrast with [homes]
-  and [<username>] both being visible). 
-
-  The presence of a share in a user's browse list does not grant the user  
-  read or write permissions on the share. Users are still able to directly 
-  access and map shares that are not visible in the browse list by using 
-  the share's UNC path. Hence, this option is of limited security utility.
+  default), the share is visible when the user lists the available shares 
+  through Windows File Explorer or through the *net view* command. If the
```
Comment (wblock) [inline userguide/sharing.rst]: The markup for a command is :command:`net view`.

File: userguide/sharing.rst [hunk 18]
```diff
@@ -1093,16 +1093,15 @@ settings:
...
-  authenticated user's name will be visible (in contrast with [homes]
-  and [<username>] both being visible). 
-
-  The presence of a share in a user's browse list does not grant the user  
-  read or write permissions on the share. Users are still able to directly 
-  access and map shares that are not visible in the browse list by using 
-  the share's UNC path. Hence, this option is of limited security utility.
+  default), the share is visible when the user lists the available shares 
+  through Windows File Explorer or through the *net view* command. If the  
+  share is flagged as a :guilabel:`home share`, unchecking this box hides the
```
Comment (wblock) [inline userguide/sharing.rst]: guilabel is for a GUI element like a button or list.  I'm not sure what "home share" represents here.  Well, a concept, but not a literal button.  I think.

File: userguide/sharing.rst [hunk 19]
```diff
@@ -1093,16 +1093,15 @@ settings:
...
-  The presence of a share in a user's browse list does not grant the user  
-  read or write permissions on the share. Users are still able to directly 
-  access and map shares that are not visible in the browse list by using 
-  the share's UNC path. Hence, this option is of limited security utility.
+  default), the share is visible when the user lists the available shares 
+  through Windows File Explorer or through the *net view* command. If the  
+  share is flagged as a :guilabel:`home share`, unchecking this box hides the 
+  *homes* share so that only the dynamically generated home share with the
+  authenticated username will be visible (in contrast with a share named
+  *homes* and the dynamically-generated home share both being visible).
```
Comment (wblock) [inline userguide/sharing.rst]: Sorry, I can't follow this explanation.  It's confusing have a share named "homes" and using the term "home share".

File: userguide/sharing.rst [hunk 20]
```diff
@@ -1287,10 +1288,10 @@ for more details.
    | xattr_tdb           | stores Extended Attributes (EAs) in a tdb file so they can be used on filesystems which do not provide support for EAs                     |
    |                     |                                                                                                                                            |
    +---------------------+--------------------------------------------------------------------------------------------------------------------------------------------+
-   | zfs_space           | correctly calculates ZFS space used by share, including space used by ZFS snapshots, quotas, and resevations. Enabled by default.          |
+   | zfs_space           | correctly calculates ZFS space used by share, including space used by ZFS snapshots, quotas, and resevations; enabled by default           |
```
Comment (wblock) [inline userguide/sharing.rst]: I think this would read better with "the share" instead of just "share".

File: userguide/sharing.rst [hunk 21]
```diff
@@ -1299,14 +1300,15 @@ for more details.
    sure to configure **all** SMB shares this way. Reboot the Mac
    client after making this change.
 
-These VFS objects do not appear in the drop-down menu and are dynamically
-configured:
+These VFS objects do not appear in the selection box:
 
 * **recycle:** moves deleted files to the recycle directory instead of
-  deleting them. Controlled by a checkbox in the share configuration.
+  deleting them. Controlled by :guilabel:`Export Recycle Bin`.
```
Comment (wblock) [inline userguide/sharing.rst]: Good!  It should still point the reader to where that checkbox is, so:

Controlled by :guilabel:`Export Recycle Bin` in the :ref:`SMB share configuration <smb_share_opts_tab>`.

File: userguide/sharing.rst [hunk 22]
```diff
@@ -1299,14 +1300,15 @@ for more details.
...
 
 * **shadow_copy2:** a more recent implementation of
-  :guilabel:`shadow_copy` with some additional features
+  :guilabel:`shadow_copy` with some additional features. *shadow_copy2*
+  and associated *shadow:* parameters are automatically added to the
```
Comment (wblock) [inline userguide/sharing.rst]: Why the colon in *shadow:*?

File: userguide/sharing.rst (hunk 19 above)
Comment (anodos325) [inline userguide/sharing.rst]: Well, the behavior is confusing as well. ;-) For instance, in my smb.conf I have the share:
[my share]
path = /mnt/dozer/share

It appears in explorer as "My Share". If I check the box "use as a homeshare", the smb4.conf is rewritten so that it has 
[homes]
path = /mnt/dozer/share

After checking the button I have two shares in windows explorer "homes" and "awalker" (my username). If I uncheck "browsable to network clients" in the share configuration, I will only have one share "awalker" because the automatically generated home share inherits the browsable parameter from the [global] section of the smb4.conf.

So the behavior is somewhat complicated and confusing for users who aren't familiar with Samba. [homes] is a special reserved section of the smb.conf file that has this unique behavior. 

So "homes" is a share. "Use as a home share" is a gui element checkbox. "Home share" refers to the share that has been flagged in the FreeNAS UI as a home share.

File: userguide/sharing.rst (hunk 22 above)
Comment (anodos325) [inline userguide/sharing.rst]: Some VFS modules in samba take additional parameters as defined in the module. FreeNAS sets the following shadow_copy2 parameters:

- shadow:snapdirseverywhere = yes
- shadow:format = <snapshot task>
- shadow:localtime = yes
- shadow:sort = desc
- shadow:snapdir = .zfs/snapshot

I tried to just write "shadow:*" but wasn't sure how the markup language would like *shadow:**

Comment (anodos325) [inline userguide/sharing.rst]: I decided to simplify this and remove reference to "shadow:*". I.e. "and associated parameters"

File: userguide/sharing.rst (hunk 20 above)
Comment (anodos325) [inline userguide/sharing.rst]: I was probably commenting in the Russian forums when I wrote that. :-) English and its silly articles.

File: userguide/sharing.rst (hunk 19 above)
Comment (anodos325) [inline userguide/sharing.rst]: Tried to clarify. Perhaps it's better to just create a separate explanation for "use as a home directory"
